## **SPAM E-mail Classification**

In [1]:
#importing libraries
import pandas as pd

In [2]:
#fetching data from given url
df=pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/spambase/spambase.data")

In [27]:
#structure of dataset

print(df.shape)

(4600, 58)


In [3]:
#viewing data
df.head()


0  0.64  0.64.1  0.1  0.32   0.2   0.3   0.4   0.5   0.6  ...  0.40  \
0  0.21  0.28    0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...  0.00   
1  0.06  0.00    0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...  0.01   
2  0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
3  0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
4  0.00  0.00    0.00  0.0  1.85  0.00  0.00  1.85  0.00  0.00  ...  0.00   

    0.41  0.42  0.778   0.43   0.44  3.756   61   278  1  
0  0.132   0.0  0.372  0.180  0.048  5.114  101  1028  1  
1  0.143   0.0  0.276  0.184  0.010  9.821  485  2259  1  
2  0.137   0.0  0.137  0.000  0.000  3.537   40   191  1  
3  0.135   0.0  0.135  0.000  0.000  3.537   40   191  1  
4  0.223   0.0  0.000  0.000  0.000  3.000   15    54  1  

[5 rows x 58 columns]

In [4]:
df.tail()

0  0.64  0.64.1  0.1  0.32   0.2  0.3  0.4  0.5  0.6  ...   0.40  \
4595  0.31   0.0    0.62  0.0  0.00  0.31  0.0  0.0  0.0  0.0  ...  0.000   
4596  0.00   0.0    0.00  0.0  0.00  0.00  0.0  0.0  0.0  0.0  ...  0.000   
4597  0.30   0.0    0.30  0.0  0.00  0.00  0.0  0.0  0.0  0.0  ...  0.102   
4598  0.96   0.0    0.00  0.0  0.32  0.00  0.0  0.0  0.0  0.0  ...  0.000   
4599  0.00   0.0    0.65  0.0  0.00  0.00  0.0  0.0  0.0  0.0  ...  0.000   

       0.41  0.42  0.778  0.43  0.44  3.756  61  278  1  
4595  0.232   0.0  0.000   0.0   0.0  1.142   3   88  0  
4596  0.000   0.0  0.353   0.0   0.0  1.555   4   14  0  
4597  0.718   0.0  0.000   0.0   0.0  1.404   6  118  0  
4598  0.057   0.0  0.000   0.0   0.0  1.147   5   78  0  
4599  0.000   0.0  0.125   0.0   0.0  1.250   5   40  0  

[5 rows x 58 columns]

In [5]:
#checking for null values

df.isnull().sum()

0         0
0.64      0
0.64.1    0
0.1       0
0.32      0
0.2       0
0.3       0
0.4       0
0.5       0
0.6       0
0.7       0
0.64.2    0
0.8       0
0.9       0
0.10      0
0.32.1    0
0.11      0
1.29      0
1.93      0
0.12      0
0.96      0
0.13      0
0.14      0
0.15      0
0.16      0
0.17      0
0.18      0
0.19      0
0.20      0
0.21      0
0.22      0
0.23      0
0.24      0
0.25      0
0.26      0
0.27      0
0.28      0
0.29      0
0.30      0
0.31      0
0.32.2    0
0.33      0
0.34      0
0.35      0
0.36      0
0.37      0
0.38      0
0.39      0
0.40      0
0.41      0
0.42      0
0.778     0
0.43      0
0.44      0
3.756     0
61        0
278       0
1         0
dtype: int64

null values are not presenet as mentioned in the documentation

In [6]:
# checking the count of label column
df["1"].value_counts()

0    2788
1    1812
Name: 1, dtype: int64

for the given data set ,1812 data is spam email and 2788 non spam

1.   List item
2.   List item



In [7]:
X = df.iloc[:,:-1]
Y = df["1"]

In [8]:
#attributes
X

0  0.64  0.64.1  0.1  0.32   0.2   0.3   0.4   0.5   0.6  ...  0.39  \
0     0.21  0.28    0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...   0.0   
1     0.06  0.00    0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...   0.0   
2     0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...   0.0   
3     0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...   0.0   
4     0.00  0.00    0.00  0.0  1.85  0.00  0.00  1.85  0.00  0.00  ...   0.0   
...    ...   ...     ...  ...   ...   ...   ...   ...   ...   ...  ...   ...   
4595  0.31  0.00    0.62  0.0  0.00  0.31  0.00  0.00  0.00  0.00  ...   0.0   
4596  0.00  0.00    0.00  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...   0.0   
4597  0.30  0.00    0.30  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...   0.0   
4598  0.96  0.00    0.00  0.0  0.32  0.00  0.00  0.00  0.00  0.00  ...   0.0   
4599  0.00  0.00    0.65  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...   0.0   

       0.40   0.41  0.42  0.778   0.43   0.44  3.756   61   278  
0     0.000  0.132   0.0  0.372  0.180  0.048  5.114  101  1028  
1     0.010  0.143   0.0  0.276  0.184  0.010  9.821  485  2259  
2     0.000  0.137   0.0  0.137  0.000  0.000  3.537   40   191  
3     0.000  0.135   0.0  0.135  0.000  0.000  3.537   40   191  
4     0.000  0.223   0.0  0.000  0.000  0.000  3.000   15    54  
...     ...    ...   ...    ...    ...    ...    ...  ...   ...  
4595  0.000  0.232   0.0  0.000  0.000  0.000  1.142    3    88  
4596  0.000  0.000   0.0  0.353  0.000  0.000  1.555    4    14  
4597  0.102  0.718   0.0  0.000  0.000  0.000  1.404    6   118  
4598  0.000  0.057   0.0  0.000  0.000  0.000  1.147    5    78  
4599  0.000  0.000   0.0  0.125  0.000  0.000  1.250    5    40  

[4600 rows x 57 columns]

some columns are having extremely high values when compared to the other, so it is better to call standard scaler for better performance

In [9]:
#target column
Y

0       1
1       1
2       1
3       1
4       1
       ..
4595    0
4596    0
4597    0
4598    0
4599    0
Name: 1, Length: 4600, dtype: int64

### **Standard Scaler**

In [10]:
#transforming values of X using standard scaler
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
sc.fit(X)
X=sc.transform(X)

#Implementing Perceptron Algorithm

In [11]:
from sklearn.linear_model import Perceptron
model = Perceptron()
model.fit(X, Y)

Perceptron()

#Evaluating model performance

In [12]:
#importing libraries for score

from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
from numpy import mean
from numpy import std

In [24]:
#crosss validation
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)

#model evaluation
scores = cross_val_score(model, X, Y, scoring='accuracy', cv=cv, n_jobs=-1)
print('Mean Accuracy: ',(mean(scores)))

Mean Accuracy:  0.8774637681159417


# Splitting the data set into two , train and test

In [14]:
#train test split with a ratio (.8,.2)


from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state=1)

In [15]:
#fitting the model with train data
model.fit(X_train,Y_train)

Perceptron()

In [16]:
#predcting the y values with x_test
Y_pred = model.predict(X_test)

In [17]:
#calculating the accuracy 

from sklearn.metrics import accuracy_score
print('Accuracy: ',accuracy_score(Y_test, Y_pred))

Accuracy:  0.8967391304347826



# Tune Perceptron Hyperparameters

##Learning Rate

In [18]:
# grid search for finding the best learning rate for the perceptron

from sklearn.model_selection import GridSearchCV

# define grid
grid = dict()
grid['eta0'] = [0.0001, 0.001, 0.01, 0.1, 1.0]
# define search
search = GridSearchCV(model, grid, scoring='accuracy', cv=cv, n_jobs=-1)
# perform the search
results = search.fit(X,Y)

results

GridSearchCV(cv=RepeatedStratifiedKFold(n_repeats=3, n_splits=10, random_state=1),
             estimator=Perceptron(), n_jobs=-1,
             param_grid={'eta0': [0.0001, 0.001, 0.01, 0.1, 1.0]},
             scoring='accuracy')

In [19]:
# evaluation
print(results.best_score_)
print(results.best_params_)


0.8960144927536232
{'eta0': 0.0001}


it is seen that model performed well with smaller learning rates .0001 , model achieved best accuracy

## Epoch Rate

In [20]:
# define grid
grid = dict()
grid['max_iter'] = [1, 10, 100, 1000, 10000]
#re defining the model best learning rate receieved .0001
re_model = Perceptron(eta0=0.0001)
# define search
search = GridSearchCV(re_model, grid, scoring='accuracy', cv=cv, n_jobs=-1)
# perform the search
results = search.fit(X,Y)
print(results.best_score_)
print( results.best_params_)

0.8960144927536232
{'max_iter': 10}


the best score and epoch rate is obtained... 
Now we have best learning rate as well as the epoch rate

#Best model with tuned parameters

In [28]:
#fitting the best model with train,test split

best_model=Perceptron(eta0=0.0001,max_iter=10)
best_model.fit(X_train,Y_train)

#predcting the y values with x_test

Y_pred = best_model.predict(X_test)

#model evaluation
print('Accuracy:',accuracy_score(Y_test, Y_pred))

Accuracy: 0.8869565217391304


In [22]:
#fitting best model

model_tuned = Perceptron(eta0=0.0001,max_iter=10)
model_tuned.fit(X,Y)

#model evaluation
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
scores = cross_val_score(model_tuned, X, Y, scoring='accuracy', cv=cv, n_jobs=-1)
print(mean(scores))

0.8960144927536232
